src/main/java/TransactionDao.java

```diff
@@ -0,0 +1,95 @@
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class TransactionDao {
+    private final Connection connection;
+
+    public TransactionDao() {
+        try {
+            connection = DriverManager.getConnection("jdbc:mysql://localhost:3306/homebudget", "root", "admin");
+        } catch (SQLException e) {
+            throw new RuntimeException();
+        }
+    }
+
+    public void close() {
+        try {
+            connection.close();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void add(Transation transation) {
+        final String sql = "INSERT INTO transaction (type, description, amount, date) VALUES (?, ?, ?, ?)";
+        try {
+            PreparedStatement statement = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
+            statement.setString(1, transation.getType());
+            statement.setString(2, transation.getDescription());
+            statement.setDouble(3, transation.getAmount());
+            statement.setString(4, transation.getDate());
+            statement.executeUpdate();
+            ResultSet generatedKeys = statement.getGeneratedKeys();
+            if (generatedKeys.next()) {
+                transation.setId(generatedKeys.getInt(1));
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public boolean update(Transation transation) {
+        final String sql = "UPDATE transaction SET type = ?, description = ?, amount = ?, date = ? where id = ?";
+        try {
+            PreparedStatement statement = connection.prepareStatement(sql);
+            statement.setString(1, transation.getType());
+            statement.setString(2, transation.getDescription());
+            statement.setDouble(3, transation.getAmount());
+            statement.setString(4, transation.getDate());
+            statement.setInt(5, transation.getId());
+            int updatedRows = statement.executeUpdate();
+            return updatedRows != 0;
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return false;
+    }
+
+    public boolean delete(int id) {
+        final String sql = "delete from transaction where id = ?";
+        try {
+            PreparedStatement statement = connection.prepareStatement(sql);
+            statement.setInt(1, id);
+            int deletedRows = statement.executeUpdate();
+            return deletedRows != 0;
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return false;
+    }
+
+    public Optional<List> getTransations(String type) {
```

Comment:
Jeżeli zwracasz kolekcję, to nie opakowuj jej dodatkowo w Optionala. Lista może być pusta, albo zawierać jakieś elementy, dodatkowe opakowywanie w Optionala, który też może być pusty, albo może zawierać obiekt jest nadmiarowe.
A jeżeli już korzystałabyś z Optionala, to należałoby też dodać parametr generyczny listy `Optional<List<Transation>>`